Mlem/API/APIClient.swift

```diff
@@ -0,0 +1,94 @@
+//
+//  APIClient.swift
+//  Mlem
+//
+//  Created by Nicholas Lawson on 04/06/2023.
+//
+
+import Foundation
+
+extension JSONDecoder {
+    static var defaultDecoder: JSONDecoder {
+        let decoder = JSONDecoder()
+        decoder.keyDecodingStrategy = .convertFromSnakeCase
+        
+        let formatter = DateFormatter()
+        
+        decoder.dateDecodingStrategy = .custom({ decoder in
+            let container = try decoder.singleValueContainer()
+            let string = try container.decode(String.self)
+            
+            formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSSSS"
+            if let date = formatter.date(from: string) {
+                return date
+            }
+            
+            formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
+            if let date = formatter.date(from: string) {
+                return date
+            }
+            
+            // TODO: is there somewhere that covers all the formats lemmy instances may return?
+            // for now we'll return the current date until we know what other formats will
+            // be encountered.
+            return .now
```

Comment:
Imo, throw here, just for correctness' sake

Reply:
Yeah, I agree, unfortunately if we `throw` then it'd take down the whole model with it, not just this property 😞 

I think the two options we have are;

1. make all `Date` properties optional by default, so in the above case we can return `nil` if we cannot decode it. This does mean the UI layer needs to handle that optionality, but would let the modelling succeed if we encounter rogue dates.
2. Fail (by throwing) when we cannot decode the date, as stated above however that would cause the entire modelling process to fail, so a bad date nested multiple layers deep in a response would fail... I'd like to avoid that.

I know you've been looking through the Rust source, is there a definitive list of date formats that the API could return? Perhaps if we ensure our formatter supports all possible date formats, but we still allow option 1 on failure it gives us the best solution... as we can be (mostly) confident all dates will format, but in an edge case we allow it to be omitted?

Reply:
There should be a definitive list, which is why I suggest throwing, because rust wouldn't allow any other variant when we have looked at all of the ones that rust could possibly return.

Reply:
Lemmy uses `chrono`, a rust library, and specific a type which says its `ISO 8601`.

Reply:
We're not parsing the "Display" variant here, I think it'll be crash-less if we implement that.

Add a throw therafter, we can recover these crashes from testflight, or via the app store's auto-reporting.

Reply:
if I'm understanding correctly that suggests the two date formats we currently attempt should be the only two returned, basically the fractional seconds may or may not be present.

Adding a `throw` will break modelling so a single rouge date in a list of posts would mean the user cannot decode/see _any_ of their posts, the app wouldn't crash in response to the `throw`, we could **make** it crash but...

I'm wondering if this is actually a problem we should solve if/when we go to the repository pattern on top of this?

We allow optionality for now, which is the safest approach in terms of modelling the API and then if we later represent these with internal app models we could force non-optionality, this would mean a single broken post etc would simply not reach the UI layer - if a user reported such a missing post we could investigate, but it sounds like the API would be the one at fault should it ever happen?

Reply:
> we could make it crash but...

Please do, if it does give us crash logs, we can make it an exception that will show the date string it crashed on

I think that if we give back a nonsensical date, we will break the app in many more subtle ways, so i think thats a bad idea

Reply:
I dont want optionality, we have these models and we should stick with them

Reply:
from the dev room:

Jo:
> actually, there may be a better way to handle this, if we dont want to crash, and if we 
>
> dont want to throw or show an error
>
> what if we return Unix timestamp 0?
> 
> Jan 1 1970 is fairly well known for that kind of stuff

Mormaer:
> yeah, I say let’s throw for now, as it is most correct 👍
> 
> if we get hit with 100s of reports after that version is out we go with the reference date (1970) and consider it more
> 
> End of the day, us not being able to parse it (assuming the API is behaving) is an edge case

Reply:
fixed in 8375791fbe626f0d5223919fdef9f474662bdabd